metadata/lib.rs

```diff
@@ -0,0 +1,580 @@
+    fn fmt(&self, f: &mut Formatter<'_>) -> DisplayResult {
+        write!(f, "{} {}", self.compiler, self.version)
+    }
+}
+
+impl Serialize for SourceCompiler {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        serializer.serialize_str(&self.to_string())
+    }
+}
+
+impl SourceCompiler {
+    pub fn new(compiler: Compiler, version: Version) -> Self {
+        SourceCompiler { compiler, version }
+    }
+}
+
+/// Compilers used to compile a smart contract.
+#[derive(Debug, Serialize)]
+pub enum Compiler {
+    RustC,
+    Solang,
+}
+
+impl Display for Compiler {
+    fn fmt(&self, f: &mut Formatter<'_>) -> DisplayResult {
+        match self {
+            Self::RustC => write!(f, "rustc"),
+            Self::Solang => write!(f, "solang"),
+        }
+    }
+}
+
+/// Metadata about a smart contract.
+#[derive(Debug, Serialize)]
+pub struct Contract {
+    name: String,
+    version: Version,
+    authors: Vec<String>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    description: Option<String>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    documentation: Option<Url>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    repository: Option<Url>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    homepage: Option<Url>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    license: Option<String>,
+}
+
+impl Contract {
+    pub fn builder() -> ContractBuilder {
+        ContractBuilder::default()
+    }
+}
+
+/// Additional user defined metadata, can be any valid json.
+#[derive(Debug, Serialize)]
+pub struct User {
+    #[serde(flatten)]
+    json: Map<String, Value>,
+}
+
+impl User {
+    /// Constructs new user metadata.
+    pub fn new(json: Map<String, Value>) -> Self {
+        User { json }
+    }
+}
+
+/// Builder for contract metadata
+#[derive(Default)]
+pub struct ContractBuilder {
+    name: Option<String>,
+    version: Option<Version>,
+    authors: Option<Vec<String>>,
+    description: Option<String>,
+    documentation: Option<Url>,
+    repository: Option<Url>,
+    homepage: Option<Url>,
+    license: Option<String>,
+}
+
+impl ContractBuilder {
+    /// Set the contract name (required)
+    pub fn name<S>(&mut self, name: S) -> &mut Self
+    where
+        S: AsRef<str>,
+    {
+        self.name = Some(name.as_ref().to_string());
+        self
+    }
+
+    /// Set the contract version (required)
+    pub fn version(&mut self, version: Version) -> &mut Self {
+        self.version = Some(version);
+        self
+    }
+
+    /// Set the contract version (required)
+    pub fn authors<I, S>(&mut self, authors: I) -> &mut Self
+    where
+        I: IntoIterator<Item = S>,
+        S: AsRef<str>,
+    {
+        self.authors = Some(
+            authors
+                .into_iter()
+                .map(|s| s.as_ref().to_string())
+                .collect(),
+        );
+        self
```

Comment:
misses asserts to check if old `authors` is empty and new `authors` is non-empty